src/signMessage.c

```diff
@@ -168,22 +176,76 @@ void handle_sign_message_parse_message(volatile unsigned int *tx) {
     }
 }
 
+static bool check_swap_validity(const SummaryItemKind_t kinds[MAX_TRANSACTION_SUMMARY_ITEMS],
+                                size_t num_summary_steps) {
+    bool amount_ok = false;
+    bool recipient_ok = false;
+    bool fee_payer_ok = false;
+    if (num_summary_steps != 3) {
+        PRINTF("3 steps expected for transaction in swap context, not %u\n", num_summary_steps);
+        return false;
+    }
+    for (size_t i = 0; i < num_summary_steps; ++i) {
+        transaction_summary_display_item(i, DisplayFlagNone | DisplayFlagLongPubkeys);
+        PRINTF("G_transaction_summary_title = %s\n", G_transaction_summary_title);
+        PRINTF("G_transaction_summary_text = %s\n", G_transaction_summary_text);
+        switch (kinds[i]) {
+            case SummaryItemAmount:
+                if (strcmp(G_transaction_summary_title, "Transfer") == 0) {
+                    amount_ok = check_swap_amount(G_transaction_summary_text);
+                } else {
+                    PRINTF("Refused field '%s'\n", G_transaction_summary_title);
+                    return false;
+                }
+                break;
+            case SummaryItemPubkey:
+                if (strcmp(G_transaction_summary_title, "Recipient") == 0) {
+                    recipient_ok = check_swap_recipient(G_transaction_summary_text);
+                } else if (strcmp(G_transaction_summary_title, "Fee payer") == 0) {
+                    fee_payer_ok = check_swap_fee_payer(G_transaction_summary_text);
+                } else {
+                    PRINTF("Refused field '%s'\n", G_transaction_summary_title);
+                    return false;
+                }
+                break;
+            default:
+                PRINTF("Refused kind '%u'\n", SummaryItemAmount);
+                return false;
+        }
+    }
+    return amount_ok && recipient_ok && fee_payer_ok;
+}
+
 void handle_sign_message_ui(volatile unsigned int *flags) {
     // Display the transaction summary
     SummaryItemKind_t summary_step_kinds[MAX_TRANSACTION_SUMMARY_ITEMS];
     size_t num_summary_steps = 0;
     if (transaction_summary_finalize(summary_step_kinds, &num_summary_steps) == 0) {
-        size_t num_flow_steps = 0;
+        // If we are in swap context, do not redisplay the message data
```

Comment:
Not sure I understand how this works, is the App called twice to sign a transaction? Once in regular and another in library mode? Could you share how I can test the swap feature? Preferably with speculos.

Reply:
I'll forward you a documentation + help on how to test and reproduce using the exchange test branch on the Discord if you are ok with this